src/mem/alloc.rs

```diff
@@ -0,0 +1,1063 @@
+            Page::Reserved(ri) => ri.encode(),
+        }
+    }
+
+    pub fn from_mem(mem: PageStorageType) -> Self {
+        let page_type = mem.0 & PAGE_TYPE_MASK;
+
+        if page_type == PAGE_TYPE_FREE {
+            Page::Free(FreeInfo::decode(mem))
+        } else if page_type == PAGE_TYPE_ALLOCATED {
+            Page::Allocated(AllocatedInfo::decode(mem))
+        } else if page_type == PAGE_TYPE_SLABPAGE {
+            Page::SlabPage(SlabPageInfo::decode(mem))
+        } else if page_type == PAGE_TYPE_COMPOUND {
+            Page::CompoundPage(CompoundInfo::decode(mem))
+        } else if page_type == PAGE_TYPE_RESERVED {
+            Page::Reserved(ReservedInfo::decode(mem))
+        } else {
+            panic!("Unknown Page Type {}", page_type);
+        }
+    }
+}
+
+struct MemoryRegion {
+    start_phys: PhysAddr,
+    start_virt: VirtAddr,
+    page_count: usize,
+    nr_pages: [usize; MAX_ORDER],
+    next_page: [usize; MAX_ORDER],
+    free_pages: [usize; MAX_ORDER],
+}
+
+impl MemoryRegion {
+    pub const fn new() -> Self {
+        MemoryRegion {
+            start_phys: PhysAddr::new(0),
+            start_virt: VirtAddr::new_truncate(0),
+            page_count: 0,
+            nr_pages: [0; MAX_ORDER],
+            next_page: [0; MAX_ORDER],
+            free_pages: [0; MAX_ORDER],
+        }
+    }
+
+    fn page_info_virt_addr(&self, pfn: usize) -> VirtAddr {
+        let size = size_of::<PageStorageType>();
+        let virt = self.start_virt;
+        virt + ((pfn as usize) * size)
+    }
+
+    fn check_pfn(&self, pfn: usize) {
+        if pfn >= self.page_count {
+            panic!("Invalid Page Number {}", pfn);
+        }
+    }
+
+    fn check_virt_addr(&self, vaddr: VirtAddr) -> bool {
+        let start: VirtAddr = self.start_virt;
+        let end: VirtAddr = self.start_virt + (self.page_count * PAGE_SIZE as usize);
+
+        vaddr >= start && vaddr < end
+    }
+
+    fn write_page_info(&self, pfn: usize, pi: Page) {
+        self.check_pfn(pfn);
+
+        let info: PageStorageType = pi.to_mem();
+        unsafe {
+            let ptr: *mut PageStorageType =
+                self.page_info_virt_addr(pfn).as_u64() as *mut PageStorageType;
+            (*ptr) = info;
+        }
+    }
+
+    fn read_page_info(&self, pfn: usize) -> Page {
+        self.check_pfn(pfn);
+
+        let virt = self.page_info_virt_addr(pfn);
+        let info: PageStorageType = PageStorageType(unsafe { *(virt.as_u64() as *const u64) });
+
+        Page::from_mem(info)
+    }
+
+    pub fn get_page_info(&self, vaddr: VirtAddr) -> Result<Page, ()> {
+        if vaddr.as_u64() == 0 || !self.check_virt_addr(vaddr) {
+            return Err(());
+        }
+
+        let pfn = (vaddr - self.start_virt) / PAGE_SIZE;
+
+        Ok(self.read_page_info(pfn as usize))
+    }
+
+    fn get_next_page(&mut self, order: usize) -> Result<usize, ()> {
+        let pfn = self.next_page[order];
```

Comment:
I know this is an internal function, but it wouldn't hurt to check order < MAX_ORDER